lib/ASTGen/Sources/ASTGen/Macros.swift

```diff
@@ -651,26 +650,16 @@ func findSyntaxNodeInSourceFile<Node: SyntaxProtocol>(
   }
 
   // If we want the outermost node, keep looking.
-  // FIXME: This is VERY SPECIFIC to handling of types. We must be able to
-  // do better.
+  // E.g. for 'foo.bar' we want the member ref expression instead of the
+  // identifier expression.
   if wantOutermost {
-    while let parentSyntax = resultSyntax.parent {
-      // Look through type compositions.
-      if let compositionElement = parentSyntax.as(CompositionTypeElementSyntax.self),
-        let compositionList = compositionElement.parent?.as(CompositionTypeElementListSyntax.self),
-        let typedParent = compositionList.parent?.as(type)
-      {
+    while
+      let parentSyntax = currentSyntax.parent,
+      parentSyntax.position == resultSyntax.position {
```

Comment:
It is a restriction that the outer nodes must stay in the same position, and it's inconsistent with the innermost node discovery. But in our use cases, this should be enough.

Reply:
Maybe we should have two separate methods like
```swift
// For macro expansion node discovery:
findTokenInSourceFile(in: exporetedSourceFile, at: sourceLoc)
  .innermostNode(ofType: AttributeSyntax.self)

// For partial ASTGen:
findTokenInSourceFile(in: exporetedSourceFile, at: sourceLoc)
  .outermostNodeAtSamePosition(ofType: ExprSyntax.self)
```